src/daft-logical-plan/src/optimization/rules/filter_null_join_key.rs

```diff
@@ -0,0 +1,233 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+use common_treenode::{Transformed, TreeNode};
+use daft_core::join::JoinType;
+use daft_dsl::optimization::conjuct;
+
+use super::OptimizerRule;
+use crate::{
+    ops::{Filter, Join},
+    LogicalPlan,
+};
+
+/// Optimization rule for filtering out nulls from join keys.
+///
+/// Inserts a filter before each side of the join to remove rows where a join key is null when it is valid to do so.
+/// This will reduce the cardinality of the tables before a join, which may improve join performance.
+#[derive(Default, Debug)]
+pub struct FilterNullJoinKey {}
+
+impl FilterNullJoinKey {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for FilterNullJoinKey {
+    fn try_optimize(&self, plan: Arc<LogicalPlan>) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        plan.transform(|node| {
+            if let LogicalPlan::Join(Join {
+                left,
+                right,
+                left_on,
+                right_on,
+                null_equals_nulls,
+                join_type,
+                ..
+            }) = node.as_ref()
+            {
+                let mut null_equals_nulls_iter = null_equals_nulls.as_ref().map_or_else(
+                    || Box::new(std::iter::repeat(false)) as Box<dyn Iterator<Item = bool>>,
+                    |x| Box::new(x.clone().into_iter()),
+                );
+
+                let (can_filter_left, can_filter_right) = match join_type {
+                    JoinType::Inner => (true, true),
+                    JoinType::Left => (false, true),
+                    JoinType::Right => (true, false),
+                    JoinType::Outer => (false, false),
+                    JoinType::Anti => (false, true),
+                    JoinType::Semi => (true, true),
+                };
+
+                let left_null_pred = if can_filter_left {
```

Comment:
how are we actually determining if the join key is nullable? AFAIK, we don't have a concept of nullable in our fields/dtypes

Reply:
like this should only push it down if the expr or colum is `null`, but we don't have a way to determine that. Maybe I'm missing something?

Reply:
This rule creates a filter that removes null rows. The join keys themselves do not have to be a null literal or type. So if the join keys are not nullable or do not have null values, this would essentially be a no-op, but if they had say a row where the value was null, it would be removed prior to the join.